### DISCUSSION TITLE
How to track supply of FA/Coin from event

### DISCUSSION THREAD

**Discussion by 0x-j** [aptos-labs/aptos-developer-discussions#41] [Questions] [answered]
Question asked by `woogieboogie.jl` in Discord.

# Background
We are currently building an ERP solution on Aptos that does Accounting/Supply Management(questions regarding burn/mint come from here) and Performance Analytics. 

# Question 1
We got the idea of comparing between withdrawEvent / depositEvent and deriving the mint/burn (mint if depositEvent - withdrawEvent > 0, burn if vice versa) from a single transaction(version) but, peeping into this version

However, here(https://aptoscan.com/transaction/488564958#events): I'm confused that this method might not work for us. 

We would expect total of four events emitted from swap transaction of A asset to B asset interacted with pancakeswap:

- withdraw A from sender, deposit A to pancakeswap
- withdraw B from pancakeswap, deposit B to sender

however we only get two events emitted which is just the results from zUSDC to Cake swap transaction, rest of the events recorded on a separate swapEvent

how should we handle this case or come up with a framework to cover such cases?

# Question 2
Regarding the sample token contract you sent us: https://explorer.aptoslabs.com/object/0x2ebb2ccac5e027a87fa0e2e5f656a3a4238d6a48d93ec9b610d570fc0aa0df12/resources?network=mainnet

could you please elaborate further on how we can catch burn/mint cases in the above example? (I'm more of a researcher than a dev... so my understanding falls short.. haha..)

and bit more questions.. regarding the example contract(object) you gave:

- we are currently collecting 0x1:Coinstore related events, but I'm guessing we should widen our coverage to 0x1:fungible_assets as FAs will be the ERC-20 equivalent token in Aptos right? what are the differences between 0x1:Coinstore vs 0x1:fungible_assets?
- what are objects (0x2ebb2ccac5e027a87fa0e2e5f656a3a4238d6a48d93ec9b610d570fc0aa0df12 on the above example)? are they just hash representations of contract addresses?

**Comment by 0x-j** [comment]
# Regarding how to track supply using events
I'll defer to @movekevin or @gregnazario 

# Regarding object and FA

> What are objects

Object is similar to contract account in Ethereum or PDA in Solana. You use it to store data (i.e. resource in move). The object you [linked](https://explorer.aptoslabs.com/object/0x2ebb2ccac5e027a87fa0e2e5f656a3a4238d6a48d93ec9b610d570fc0aa0df12/resources?network=mainnet) stores metadata of a fungible asset `CELLANA`, it also stores the capabilities (mint ref, burn ref, etc) of the FA, meaning whoever is the owner of the object can mint and burn the FA. In some sense, this object is similar to the state of an ERC-20 contract. You can learn more about Object [here](https://aptos.dev/standards/aptos-object).

> but I'm guessing we should widen our coverage to 0x1:fungible_assets as FAs will be the ERC-20 equivalent token in Aptos 

Correct, fungible asset is the successor to coin, FA builds on top of Object, providing better performance and extendability. You can learn more about FA [here](https://aptos.dev/standards/fungible-asset/) and how to create a FA [here](https://aptos.dev/tutorials/your-first-fungible-asset).

**Comment by gregnazario** [comment, marked as answer]
## Question 1
Coins are be replaced by Fungible Assets for specifically what you're talking about.

It's really hard to track coins, because the events only come when they are in a CoinStore.  But, when they're in a DEX or elsewhere, those events are not available.  So, in that case you only get as you said.

Fungible assets will be easier, as you'll always have a withdraw and deposit event as the fungible assets must be stored in a fungible store, and we're thinking about adding mint and burn events as well.

The cases would increase based on the platforms you support for Coin (e.g. liquidswap, thala, etc.)

## Question 2

There's a lot here:
1. Fungible Assets are headed towards replacing coins, for some of the reasons you say. Both are ERC-20 similar
2. Objects are a collection of resources at an address, with an owner.  This allows for parallelism and simple grouping of assets.  https://aptos.dev/move/move-on-aptos/objects/ is a more detailed usage guide.
3. If I were you, I would index both, but Fungible assets are easier to index because:
  1. All fungible assets are tracked, you will have withdraw and deposit events
  2. All deposits - withdrawals have to add up to 0, otherwise it's a mint or burn, depending on if it's positive or negative.

**Reply by woogieboogie-jl** [reply]
Thx Greg. This clarifies the most of misunderstandings we had.
